From bc2b78339517cb4b1814f1497a54dfc6f825e448 Mon Sep 17 00:00:00 2001
From: Liu Ying <victor.liu@nxp.com>
Date: Thu, 11 Oct 2018 16:45:33 +0800
Subject: [PATCH 4815/5242] MLK-19887 drm/imx: dpu: crtc: Workaround half
 screen issue when pc is used

commit  38b8726c5bfbaaff4a1dba3e93197e0f996ea8c3 from
https://source.codeaurora.org/external/imx/linux-imx.git

When pixel combiner is used, it turns out we have to turn TCON from
bypass mode to operation mode and then wait for Framegen frame counter
moving for the two display streams, otherwise, the display pipeline
cannot be setup correctly sometimes(one of the two display streams
would be broken).

Signed-off-by: Liu Ying <victor.liu@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/drm/imx/dpu/dpu-crtc.c |    9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/imx/dpu/dpu-crtc.c b/drivers/gpu/drm/imx/dpu/dpu-crtc.c
index e15878f..e3ad4d8 100644
--- a/drivers/gpu/drm/imx/dpu/dpu-crtc.c
+++ b/drivers/gpu/drm/imx/dpu/dpu-crtc.c
@@ -215,12 +215,17 @@ static void dpu_crtc_atomic_enable(struct drm_crtc *crtc,
 	 * TKT320590:
 	 * Turn TCON into operation mode later after the first dumb frame is
 	 * generated by DPU.  This makes DPR/PRG be able to evade the frame.
+	 * However, if pixel combiner is used, it turns out we have to set
+	 * the TCON into operation mode first and then wait for Framegen
+	 * frame counter moving, otherwise, the display pipeline cannot be
+	 * setup correctly sometimes(one of the two display streams would
+	 * be broken).  This is a mysterious issue.
 	 */
 	if (dcstate->use_pc) {
-		framegen_wait_for_frame_counter_moving(dpu_crtc->m_fg);
 		tcon_set_operation_mode(dpu_crtc->m_tcon);
-		framegen_wait_for_frame_counter_moving(dpu_crtc->s_fg);
 		tcon_set_operation_mode(dpu_crtc->s_tcon);
+		framegen_wait_for_frame_counter_moving(dpu_crtc->m_fg);
+		framegen_wait_for_frame_counter_moving(dpu_crtc->s_fg);
 
 		framegen_wait_for_secondary_syncup(dpu_crtc->m_fg);
 		framegen_wait_for_secondary_syncup(dpu_crtc->s_fg);
-- 
1.7.9.5

